backend/src/nodes/impl/dithering.py

```diff
@@ -0,0 +1,427 @@
+
+
+def find_closest_uniform_color(value: float, num_colors: int) -> float:
+    return np.floor(value * (num_colors - 1) + 0.5) / (num_colors - 1)
+
+
+def one_channel_uniform_quantize(image: np.ndarray, num_colors: int) -> np.ndarray:
+    out_image = uniform_quantize_image(
+        dtype_to_float(image), num_colors=num_colors
+    )
+    return float_to_dtype(out_image, image.dtype)
+
+
+def one_channel_ordered_dither(image: np.ndarray, threshold_map: ThresholdMap, num_colors: int) -> np.ndarray:
+    """
+    Apply an ordered dithering algorithm to the input greyscale image.  The output will be dithered and
+    quantized to the given number of evenly-spaced values.
+
+    The output will be the same shape and dtype as the input.
+    """
+
+    tm = get_threshold_map(image.shape, threshold_map=threshold_map)
+    out_image = uniform_quantize_image(
+        dtype_to_float(image) + tm, num_colors=num_colors
+    )
+    return float_to_dtype(out_image, image.dtype)
+
+
+def one_channel_error_diffusion(image: np.ndarray, num_colors: int,
+                                error_diffusion_map: ERROR_DIFFUSION_MAP_TYPE) -> np.ndarray:
+    output_image = dtype_to_float(image)
+    edm = get_error_diffusion_map(error_diffusion_map)
+    for j in range(output_image.shape[1]):
+        for i in range(output_image.shape[0]):
+            pixel = output_image[i, j]
+            output_image[i, j] = find_closest_uniform_color(pixel, num_colors)
+            error = pixel - output_image[i, j]
+            for (di, dj), coefficient in edm.items():
+                if i + di >= output_image.shape[0] or j + dj >= output_image.shape[1]: continue
+                output_image[i + di, j + dj] += error * coefficient
+    return float_to_dtype(output_image, image.dtype)
+
+
+def apply_to_all_channels(one_channel_filter, image: np.ndarray, *args, **kwargs) -> np.ndarray:
+    if image.ndim == 2:
+        return one_channel_filter(image, *args, **kwargs)
+    output_image = np.stack(
+        [
+            one_channel_filter(image[:, :, channel], *args, **kwargs)
+            for channel in range(image.shape[2])
+        ],
+        axis=2,
+    )
+    return output_image
+
+
+def uniform_quantize(image: np.ndarray, num_colors: int) -> np.ndarray:
+    return apply_to_all_channels(one_channel_uniform_quantize, image, num_colors=num_colors)
+
+
+def ordered_dither(image: np.ndarray, threshold_map: ThresholdMap, num_colors: int) -> np.ndarray:
+    return apply_to_all_channels(one_channel_ordered_dither,
+                                 image, threshold_map=threshold_map, num_colors=num_colors)
+
+
+def error_diffusion_dither(image: np.ndarray, error_diffusion_map: ErrorDiffusionMap, num_colors: int) -> np.ndarray:
+    return apply_to_all_channels(one_channel_error_diffusion,
+                                 image, num_colors=num_colors, error_diffusion_map=error_diffusion_map)
+
+
+# TODO handle palettes
```

Comment:
That's a lot of code, and it seems like more will come. Maybe we could make one file per dithering method?

Reply:
Yes, good idea.  I'll break it up